gorm-tools/src/main/groovy/gorm/tools/repository/BulkableRepo.groovy

```diff
@@ -0,0 +1,162 @@
+        Class<J> jobClass = (Class<J>)GenericTypeResolver.resolveTypeArguments(getClass(), BulkableRepo)[1]
+        J job = jobClass.newInstance() as J
+
+        // XXX We can do something similar for error count
+        AtomicInteger count = new AtomicInteger(0)
+
+        // for error handling -- based on `onError` from args commit success and report the failure or fail them all
+        //  also use errorThreshold, for example if it failed on more than 10 records we stop and rollback everything
+        //@jdabal - for parallel async, we can not rollback batches which are already commited
+        //as each batch would be in its own transaction.
+        List bulkResult = []
+        if(parallelProcessingEnabled) {
+            //run the batch parallel in batched transactions
+            asyncSupport.parallel([batchSize:batchSize], dataList.collate(batchSize)) { List<Map> batch ->
+                // returns entity that was created, but into Job object we only want id and sourceid.
+                // id:123, source.sourceId, -  @jdabal assume that every domain which is bulk imported haas sourceId field ?
+                // store results in resultList,but id and sourceId only for succesfully created records
+                List results = doBulkCreate(batch, args)
+                if(results) bulkResult.addAll transformResults(results)
+            }
+        } else {
+            List results = doBulkCreate(dataList, args)
+            if(results) bulkResult.addAll transformResults(results)
+        }
+
+        count.getAndAdd(bulkResult.size())
+
+        job['results'] = Jsonify.render(bulkResult).jsonText.bytes
+        job.persist()
+
+        //XXX  https://github.com/9ci/domain9/issues/331 assign jobId on each record created.
+        // Special handling for arTran - we will have ArTranJob (jobId, arTranId). For all others we would
+        //have JobLink (jobId,entityId, entityName)
+        return job
+    }
+
+    List<Results> doBulkCreate(List<Map> dataList, Map args = [:]){
+        List<Results> resultList = [] as List<Results>
+        for (Map item : dataList) {
+            D entity = (D)getEntityClass().newInstance()
+            Results r
+            try {
+                //need to do bindAndCreate - so that even if create fails, we can get source.sourceId if it is created
+                bindAndCreate(entity, item, args)
+                r = Results.OK().id(entity["id"] as Long)
+            } catch(Exception e) {
+                r = Results.error(e)
+            }
+            String sourceId = getSourceId(entity, item)
+            if(sourceId) {
+                r.meta['sourceId'] = sourceId
+            }
+            resultList.add r
+        }
+        return resultList
+    }
+
+    private String getSourceId(D entity, Map item) {
+        String sourceId = null
+        //if the domain has source.sourceId return it
+        if(entity && entity.hasProperty("source") && entity['source']) {
```

Comment:
should be using traits if we are going to do this. we could be doing it based on a config too